### PR TITLE
V2beta1 Updates

### PR DESCRIPTION
**Pull request by MarkPAndrews**
### Summary

Migrates the v2beta1 API spec from a YAML file to the canonical JSON source at `https://api.neo4j.io/v2beta1/spec.json`, aligns the local resource folder structure to match that URL path, and implements all new endpoints and error responses introduced in the updated spec.

---

### Spec management changes

- **New spec location** — spec.json mirrors the upstream URL path (`/v2beta1/spec.json`)
- **Download/validation scripts** are now format-aware: v2beta1 is treated as JSON while v1 and v1beta5 remain YAML
- **Absolute URL support** — spec URLs from `swagger-initializer.js` that are already fully-qualified are passed through unchanged instead of being prefixed with the base path
- **Migration fallback** — the validator transparently falls back to the legacy aura_api_spec_v2beta1.yaml if the new JSON file is not yet present, using semantic (canonical JSON) hashing for cross-format comparison
- **Fallback hardcoded specs** in `_get_fallback_specs()` updated to match new paths and formats

---

### v1 spec changes

- Added `HTTP 420 Unsupported Action` response to the instance overwrite endpoint (attempting to overwrite a free-tier instance)

---

### New error handling

Two new exception classes have been added, following the same pattern as existing exceptions:

| Exception | HTTP Status | Meaning |
|-----------|-------------|---------|
| `AuraApiUnsupportedActionException` | 420 | Operation is unsupported for this instance tier |
| `AuraApiValidationException` | 422 | Request body failed server-side validation, a resource limit was reached, or the operation is unsupported for the target resource type |

Both are exported from the package root and have unit test coverage.

---

#### Organization invites
| Method | Path |
|--------|------|
| `GET` / `POST` | `/organizations/{org_id}/invites` |
| `DELETE` | `/organizations/{org_id}/invites/{invite_id}` |

#### User management additions
| Method | Path |
|--------|------|
| `PATCH` | `/organizations/{org_id}/users/{user_id}` |
| `POST` | `/organizations/{org_id}/projects/{project_id}/users/{user_id}` |

#### Graph analytics sessions
| Method | Path |
|--------|------|
| `GET` | `/organizations/{org_id}/graph-analytics/sessions` |
| `GET` / `POST` | `/organizations/{org_id}/projects/{project_id}/graph-analytics/sessions` |
| `POST` | `/organizations/{org_id}/projects/{project_id}/graph-analytics/sessions/sizing` |
| `GET` / `DELETE` | `/organizations/{org_id}/projects/{project_id}/graph-analytics/sessions/{session_id}` |

#### Project-scoped instances and databases
| Method | Path |
|--------|------|
| `GET` / `POST` | `/organizations/{org_id}/projects/{project_id}/instances` |
| `GET` / `DELETE` | `/organizations/{org_id}/projects/{project_id}/instances/{instance_id}` |
| `GET` / `POST` | `/organizations/{org_id}/projects/{project_id}/instances/{instance_id}/databases` |
| `GET` / `DELETE` | `/organizations/{org_id}/projects/{project_id}/instances/{instance_id}/databases/{database_id}` |
| `GET` / `POST` | `/organizations/{org_id}/projects/{project_id}/instances/{instance_id}/databases/{database_id}/backups` |
| `GET` | `/organizations/{org_id}/projects/{project_id}/instances/{instance_id}/databases/{database_id}/backups/{backup_id}` |
| `POST` | `/organizations/{org_id}/projects/{project_id}/instances/{instance_id}/databases/{database_id}/restore` |

---

### New models

| Model | Purpose |
|-------|---------|
| `AuraApiUnsupportedActionException` | HTTP 420 client error |
| `AuraApiValidationException` | HTTP 422 client error |
| `OrganizationInvite`, `OrganizationInvitesResponse`, `OrganizationInviteResponse` | Invite read models |
| `CreateOrganizationInviteRequest` | Invite creation payload |
| `AddProjectUserRequest` | Project user add payload |
| `GDSError`, `SessionCreatedBy`, `SessionResponse`, `SessionsResponse`, `SessionEnvelope` | Graph analytics session read models |
| `SessionSizeResponse`, `SessionSizeEnvelope` | Session size estimation response |
| `CreateGraphAnalyticsSessionRequest`, `SessionSizingRequest` | Session request payloads |
| `ProjectInstanceSummary`, `ProjectInstanceDetails`, `ProjectInstancesResponse`, `ProjectInstanceResponse` | Project instance models |
| `CreateProjectInstanceRequest` | Instance creation payload |
| `ProjectDatabaseSummary`, `ProjectDatabasesResponse`, `ProjectDatabaseResponse` | Database models |
| `ProjectDatabaseBackup`, `ProjectDatabaseBackupsResponse`, `ProjectDatabaseBackupResponse`, `CreateProjectDatabaseBackupResponse` | Backup models |
| `CreateProjectDatabaseRequest`, `RestoreProjectDatabaseRequest` | Database/restore payloads |
| `OpenRequestModel` | Forward-compatible base class for request models (`extra="allow"`) |

---

### Tests

- Unit test coverage added for all new v2beta1 endpoints in test_client_v2.py
- Unit tests added for HTTP 420 → `AuraApiUnsupportedActionException` and HTTP 422 → `AuraApiValidationException` in test_client.py
- All 63 unit tests pass; `black`, `isort`, and `ruff` checks clean

---

### Breaking changes

None. Existing v1, v1beta5, and v2beta1 client methods are unchanged.

### CODE REVIEW

**Review by angelosantos4 — APPROVED** [review]

File: neo4j_aura_sdk/client.py [hunk 1]
```diff
@@ -768,6 +825,184 @@ async def list_organizations(self):
...
+
+        Returns: List of OrganizationUser objects.
+        """
+        self._ensure_api_is_v2()
+        items = await self._get(
+            f"organizations/{organizationId}/users",
+            model=None,
+            api_version="v2beta1",
+        )
+        return [OrganizationUser(**item) for item in items] if items else []
```
[the remark this answers is not in the document]
Comment (MarkPAndrews) [reply neo4j_aura_sdk/client.py]: Fixed. Updated client parsing to read the v2beta1 envelope shape { data: [...] } and keep compatibility with list payloads. Added/updated unit coverage for the envelope response.

File: neo4j_aura_sdk/client.py [hunk 2]
```diff
@@ -768,6 +825,184 @@ async def list_organizations(self):
...
+        The current API does not require a request body for this operation.
+        Returns the updated OrganizationUserDetails.
+        """
+        self._ensure_api_is_v2()
+        result = await self._patch(
+            f"organizations/{organizationId}/users/{userId}",
+            body=JobIdEnvelope(),
+            model=None,
+            api_version="v2beta1",
+        )
```
[the remark this answers is not in the document]
Comment (MarkPAndrews) [reply neo4j_aura_sdk/client.py:892]: Fixed. Removed the request body from patch_organization_user and made PATCH helper body optional. Added test assertion that request body is empty

File: neo4j_aura_sdk/models.py [hunk 3]
```diff
@@ -587,3 +606,415 @@ class UsageResponse(BaseModel):
...
+
+        # Backward compatibility: older payloads used `config`.
+        if self.parameters is None and self.config is not None:
+            self.parameters = self.config
+            self.config = None
+
+        if self.parameters is None:
+            raise ValueError(
+                "similaritySearch tools require `parameters` with provider, model, and index"
+            )
```
[the remark this answers is not in the document]
Comment (MarkPAndrews) [reply neo4j_aura_sdk/models.py]: Fixed. similaritySearch validation now accepts either parameters or config and preserves config for serialization. Added tests to verify config is emitted as config (not parameters).

File: neo4j_aura_sdk/models.py [hunk 4]
```diff
@@ -587,3 +606,415 @@ class UsageResponse(BaseModel):
...
+    created_by: Optional[str] = None
+    project_id: Optional[str] = None
+    organization_id: Optional[str] = None
+    system_prompt: Optional[str] = None
+    dbid: Optional[str] = None
+    created_at: Optional[str] = None
+    updated_at: Optional[str] = None
+    is_private: Optional[bool] = None
+    is_mcp_enabled: Optional[bool] = None
+    tools: Optional[List[AgentTool]] = None
```
[the remark this answers is not in the document]
Comment (MarkPAndrews) [reply neo4j_aura_sdk/models.py]: Fixed. Introduced a lightweight AgentToolSummary for list responses and switched ListAgentResponse.tools to that type, so summary tools (name/type only) parse correctly.

**Review by MarkPAndrews — COMMENTED** [review]
Implemented suggested changes

File: neo4j_aura_sdk/client.py [hunk 5]
```diff
@@ -768,6 +825,184 @@ async def list_organizations(self):
...
+        organizationId: str,
+        projectId: str,
+        userId: str,
+        details: AddProjectUserRequest | None = None,
+    ):
+        """Add a user to a project (v2beta1)."""
+        self._ensure_api_is_v2()
+        result = await self._post(
+            f"organizations/{organizationId}/projects/{projectId}/users/{userId}",
+            body=details or JobIdEnvelope(),
```
Comment (ccloes) [inline neo4j_aura_sdk/client.py]: When `details` is `None`, this sends a `JobIdEnvelope()` as the body, which serializes to `{"data": null}` (verified locally). That's an unrelated model and a semantically wrong payload for a project-user endpoint. Suggest either sending no body (`body=details`) if the endpoint accepts an empty POST, or defaulting to `AddProjectUserRequest()` so the payload shape matches the resource.

File: neo4j_aura_sdk/client.py [hunk 6]
```diff
@@ -768,6 +825,184 @@ async def list_organizations(self):
...
+
+        Returns: OrganizationUserDetails including project memberships.
+        """
+        self._ensure_api_is_v2()
+        result = await self._get(
+            f"organizations/{organizationId}/users/{userId}",
+            model=None,
+            api_version="v2beta1",
+        )
+        return OrganizationUserDetails(**result["data"])
```
Comment (ccloes) [inline neo4j_aura_sdk/client.py]: `result["data"]` assumes a non-empty body with a `data` key. `_get` returns `None` on a 204/empty response and returns the raw dict when `model=None`, so this raises `TypeError: 'NoneType' object is not subscriptable` (empty body) or `KeyError: 'data'` (unwrapped body). The sibling methods `add_project_user`/`patch_organization_user` already guard with `result.get("data", result) if isinstance(result, dict) else result` — applying that same guard here keeps it consistent.

File: neo4j_aura_sdk/client.py [hunk 7]
```diff
@@ -768,6 +825,184 @@ async def list_organizations(self):
...
+        Returns: ProjectUser with updated role.
+        """
+        self._ensure_api_is_v2()
+        result = await self._patch(
+            f"organizations/{organizationId}/projects/{projectId}/users/{userId}",
+            body=details,
+            model=None,
+            api_version="v2beta1",
+        )
+        return ProjectUser(**result["data"])
```
Comment (ccloes) [inline neo4j_aura_sdk/client.py]: Same issue as `get_organization_user`: `result["data"]` will crash on a 204/empty or unwrapped response. Please apply the same `result.get("data", result) if isinstance(result, dict) else result` guard used in the other new user methods for consistency.

File: neo4j_aura_sdk/client.py [hunk 8]
```diff
@@ -950,6 +1200,278 @@ async def cancel_import_job(self, organizationId: str, projectId: str, jobId: st
...
+            api_version="v2beta1",
+            params=params if params else None,
+        )
+
+    async def list_project_graph_analytics_sessions(
+        self, organizationId: str, projectId: str
+    ):
+        """List graph analytics sessions for a project (v2beta1)."""
+        self._ensure_api_is_v2()
+        return await self._get(
```
Comment (ccloes) [inline neo4j_aura_sdk/client.py]: Return-shape inconsistency across the new methods: this returns the raw `SessionsResponse` envelope, while `list_organization_invites` returns `result.data or []` and `list_organization_users`/`list_project_users` hand-roll `[Model(**item) ...]`. Structurally identical list endpoints now hand callers three different return types. Worth aligning on one convention before this becomes public API surface (not blocking).

File: neo4j_aura_sdk/client.py [hunk 9]
```diff
@@ -792,14 +1029,23 @@ async def get_billing_usage(
             params["page_token"] = page_token
         if page_limit:
             params["page_limit"] = page_limit
+        if project_id:
+            params["project_id"] = project_id
```
Comment (ccloes) [inline neo4j_aura_sdk/client.py:1037]: This `project_id` list parameter is new. httpx serializes a list into repeated query keys (`project_id=a&project_id=b`). Please confirm that's the encoding the spec expects (vs. CSV) and add a test — none of the current tests exercise the list-valued param path.

**Review by ccloes — COMMENTED** [review]
A few correctness notes on the new v2beta1 methods. All comments below are on code newly added (or newly parameterized) in this PR; I've left pre-existing style issues out of scope. Nice work overall — 65/65 unit tests pass locally and the 420/422 handling + `_put` helper look clean.

File: neo4j_aura_sdk/client.py (hunk 6 above)
Comment (MarkPAndrews) [reply neo4j_aura_sdk/client.py]: Good catch.  Fixed.

File: neo4j_aura_sdk/client.py (hunk 7 above)
Comment (MarkPAndrews) [reply neo4j_aura_sdk/client.py]: fixed

File: neo4j_aura_sdk/client.py (hunk 5 above)
Comment (MarkPAndrews) [reply neo4j_aura_sdk/client.py]: Update to just send details.  The model can be empty so that is the easiest fix,

File: neo4j_aura_sdk/client.py (hunk 9 above)
Comment (MarkPAndrews) [reply neo4j_aura_sdk/client.py:1037]: The old YAML spec has the parameters, but the new JSON spec omits them.  The test has been tightened up to ensure repeated keys.

File: neo4j_aura_sdk/client.py (hunk 8 above)
Comment (MarkPAndrews) [reply neo4j_aura_sdk/client.py]: From Copilot: Normalized the new `list_*` client methods to all return plain lists instead of a mix of envelope models and lists. That brings `list_organizations`, `list_organization_projects`, `list_organization_graph_analytics_sessions`, `list_project_graph_analytics_sessions`, `list_project_instances`, `list_project_instance_databases`, `list_project_database_backups`, and `list_deployments` in line with the existing `list_organization_invites`, `list_organization_users`, and `list_project_users` behavior, so callers no longer have to guess whether to access `.data` or iterate the result directly. I also updated the unit tests to lock in the new return shape.

**Review by ccloes — APPROVED** [review]

**Comment by ccloes** [comment]
Feel free to merge this... I will leave this to you to merge.
